alab_management/resource_manager/resource_manager.py

```diff
@@ -0,0 +1,207 @@
+"""
+
+import time
+from datetime import datetime
+from typing import Any, cast
+
+import dill
+from bson import ObjectId
+
+from alab_management.device_view.device_view import DeviceView
+from alab_management.logger import DBLogger
+from alab_management.resource_manager.enums import _EXTRA_REQUEST
+from alab_management.resource_manager.resource_requester import (
+    RequestMixin,
+    RequestStatus,
+)
+from alab_management.sample_view.sample import SamplePosition
+from alab_management.sample_view.sample_view import SamplePositionRequest, SampleView
+from alab_management.task_view import TaskView
+from alab_management.task_view.task_enums import TaskStatus
+from alab_management.utils.data_objects import get_collection
+from alab_management.utils.module_ops import load_definition
+
+
+class ResourceManager(RequestMixin):
+    """
+    TaskManager will.
+
+    (1) find all the ready tasks and submit them,
+    (2) handle all the resource requests
+    """
+
+    def __init__(self):
+        load_definition()
+        self.task_view = TaskView()
+        self.sample_view = SampleView()
+        self.device_view = DeviceView()
+        self._request_collection = get_collection("requests")
+
+        self.logger = DBLogger(task_id=None)
+        super().__init__()
+        time.sleep(1)  # allow some time for other modules to launch
+
+    def run(self):
+        """Start the loop."""
+        while True:
+            self._loop()
+            # time.sleep(1)
+
+    def _loop(self):
+        self.handle_released_resources()
+        self.handle_requested_resources()
+
+    def handle_released_resources(self):
+        """Release the resources."""
+        for request in self.get_requests_by_status(RequestStatus.NEED_RELEASE):
+            devices = request["assigned_devices"]
+            sample_positions = request["assigned_sample_positions"]
+            self._release_devices(devices)
+            self._release_sample_positions(sample_positions)
+            self.update_request_status(
+                request_id=request["_id"], status=RequestStatus.RELEASED
+            )
+
+    def handle_requested_resources(self):
+        """
+        Check if there are any requests that are in PENDING status. If so,
+        try to assign the resources to it.
+        """
+        requests = list(self.get_requests_by_status(RequestStatus.PENDING))
+        # prioritize the oldest requests at the highest priority value
+        requests.sort(key=lambda x: x["submitted_at"])
+        requests.sort(key=lambda x: x["priority"], reverse=True)
+        for request in requests:
+            self._handle_requested_resources(request)
+
+    def _handle_requested_resources(self, request_entry: dict[str, Any]):
+        try:
+            resource_request = request_entry["request"]
+            task_id = request_entry["task_id"]
+
+            task_status = self.task_view.get_status(task_id=task_id)
+            if task_status != TaskStatus.REQUESTING_RESOURCES:
```

Comment:
are we 100% sure that when it is not request resources means that is error or being canceled somewhere else?
I am just trying to understand if there is another case.

Reply:
Yes. Because this code is basically for the racing condition. The `_handle_requested_resources` can only be called on the task that is in `REQUESTING_RESOURCES` state. If it is no longer in this state, it means something bad happens and should be stopped.